RequestTileEntityInteraction hook and handling.

Patrikkk opened the pull request:
Implemented the OnRequestTileEntityInteraction even hook in GetDataHandler.
Created RequestTileEntityInteractionHandler which checks for building permissions when the player is attempting to open a display doll (Mannequin). This now prevents players from opening a mannequin if they have no building permissions at the position of the Mannequin.

Patrikkk committed:
GetDataHandler - Code formatting. Modify order of Handler to respect PacketType value.

Patrikkk committed:
GetDataHandlers - Add RequestTileEntityInteraction hook.

Patrikkk committed:
Add RequestTileEntityInteractionHandler. Check Mannequin modif perms.

Users can no longer open a Mannequin if they do not have building permissions at the position of the Mannequin.
(Mannequins work as a chest since 1.4)

Patrikkk committed:
Update CHANGELOG.md

Patrikkk committed:
Changelog - Grammar fixes

Patrikkk commented:
Referencing issue #2018

Patrikkk committed:
Add building permission checks for Hat Rack modification.

Keeping this code format to have the code be friendly with us in future updates, as well as display proper rejection message to the players.

Patrikkk committed:
Update CHANGELOG.md

Patrikkk committed:
Merge branch 'general-devel' into addtileentityinteraction

QuiCM commented on TShockAPI/Handlers/RequestTileEntityInteractionHandler.cs:
```diff
@@ -0,0 +1,48 @@
+			if (args.TileEntityID != -1)
+			{
+				TileEntity tileEntity;
+				if (TileEntity.ByID.TryGetValue(args.TileEntityID, out tileEntity))
```
Can this ever set tileEntity to `null`?

bartico6 replied:
This is a packet handler, so if the client sends an invalid ID on purpose (or sends one right as one was removed) this will return false.

Patrikkk replied:
TileEntity.ByID is a dictionary of int, TileEntity, and the dictionary does not store any nulls to the ID.s 
![](https://puu.sh/FYOOj/ab5e532b07.png)
So this block only executes if the TryGet returns true and it can get the entity by ID, and it will not return a null object.
In my next PR regarding DollItemSync, I am passing the tileentity object. 
Should I modify this PR and pass the tile entity object, instead of just the ID, so it gives better usability for Devs when they work with this hook?

QuiCM replied:
> 
> Should I modify this PR and pass the tile entity object, instead of just the ID, so it gives better usability for Devs when they work with this hook?

That sounds good to me

Patrikkk committed:
TileEntityInteraction - Pass TileEntity object instead of ID in args.

Patrikkk committed:
Merge branch 'general-devel' into addtileentityinteraction

QuiCM commented on TShockAPI/Handlers/RequestTileEntityInteractionHandler.cs:
```diff
@@ -0,0 +1,41 @@
+	{
+		public void OnReceive(object sender, RequestTileEntityInteractionEventArgs args)
+		{
+			if (args.TileEntity is TEHatRack && !args.Player.HasBuildPermission(args.TileEntity.Position.X, args.TileEntity.Position.Y, false))
```
Might be worth using `args.Player.HasBuildPermissionForTileObject` here, as it will check all tiles in the entity for build permission

Patrikkk replied:
Do you know what is the difference between the two checks? If not, I can go have a look.

Patrikkk replied:
Ah, TileObject requires width and height. This data is not in a TileEntity object, would need to be cast as a TEHatRack or TEDisplayDoll.

QuiCM commented on TShockAPI/Handlers/RequestTileEntityInteractionHandler.cs:
```diff
@@ -0,0 +1,41 @@
+		{
+			if (args.TileEntity is TEHatRack && !args.Player.HasBuildPermission(args.TileEntity.Position.X, args.TileEntity.Position.Y, false))
+			{
+				args.Player.SendErrorMessage("You do not have permission to modify a Hat Rack in a protected area!");
```
These if statements all perform the same logic.
Might be nice to separate it such that the if statements determine what text to send and then perform the logic outside of it.
Eg
```csharp
string type = "Tile Entity";
if (args.TileEntity is TEHatRack)
{
    type = "Hat Rack";
}
// etc

args.Player.SendErrorMessage($"You do not have permission to modify this {type}.");
TShock.Log.ConsoleDebug($"RequestTileEntityInteractionHandler: Rejected packet due to lack of building permissions - From {args.Player.Name} | Position X:{args.TileEntity.Position.X} Y:{args.TileEntity.Position.Y}, TileEntity type: {args.TileEntity.type}, Tile type: {Main.tile[args.TileEntity.Position.X, args.TileEntity.Position.Y].type}");
args.Handled = true;
return;
```

Patrikkk replied:
What's your reasoning for that code approach? 
I'm assuming is to have one block of logic that (the output messages) with a variable to tend to, rather than having the same code with a hardcoded variant.
It is rather only visuals/organization, than performance because it would not change much at all. (an extra string being instantiated is minuscule) 
But yeah I don't mind to change it to that at all. Just simply curious of your way of thinking:)

QuiCM replied:
Mostly just clarity. We have many spots in TShock where we have the same logic run in multiple if/else blocks with slightly different messages and it ends up being hard to maintain/adjust/extend

Patrikkk replied:
Yeah, I was thinking the same once you've pointed it out. Thanks!:)

Patrikkk replied:
If you check the other comment above, to use HasBuildPermissionForTileObject I would need to cast the object as TEHatRack or TEDisplayDoll.  This means I would need to have 

```c#
if (args.TileEntity is TEHatRack && !args.Player.HasBuildPermissionForTileObject(args.TileEntity.Position.X, args.TileEntity.Position.Y, ((TEHatRack)args.TileEntity).entityTileWidth, ((TEHatRack)args.TileEntity).entityTileHeight,  false))
```

And keep the same logic structure.

Before looking into ForTileObject perm check I made this edit
![](https://puu.sh/G0jVG/12de92cfa2.png)
![](https://puu.sh/G0kav/bf6eb27ce3.png)
![](https://puu.sh/G0kb4/d5bd4a56e0.png)

Patrikkk committed:
Merge branch 'general-devel' into addtileentityinteraction

hakusaro committed:
Merge branch 'general-devel' into addtileentityinteraction

hakusaro commented:
FYI: I've updated your changelog entry to match renormalized changelog changes on the changelog.

Patrikkk committed:
RequestTileEntityInteractionHandler - Use object size for building perm check.

This commit will modify the RequestTileEntityInteractionHandler to use HasBuildPermissionForTileObject when checking for building permissions for Hat Rack and Display Doll to give an accurate response wether or not any part of the object has an overlapping protected region.

Patrikkk commented:
I've now implemented a change that I see the best fit for allowing HasBuildPermissionForTileObject to be used for an accurate check over a Hat Rack and Mannequin. If @QuiCM is okay with this approach, feel free to merge.
I'm open for suggestions for better code handling.

QuiCM commented:
> 
> 
> I've now implemented a change that I see the best fit for allowing HasBuildPermissionForTileObject to be used for an accurate check over a Hat Rack and Mannequin. If @QuiCM is okay with this approach, feel free to merge.
> I'm open for suggestions for better code handling.

For the moment I think this is the best solution. Ideally I'd like to be able to condense the if/else blocks but from earlier discussion and experimentation its more trouble than its worth.
LGTM
